src/services/factories/tokenList.ts

```diff
@@ -130,19 +132,35 @@ export function getTokensFactory(
     return new Map(tokenAddressIdPairs)
   }
 
+  async function fetchTcrAddresses(networkId: number): Promise<Set<string>> {
+    return new Set(await tcrApi.getTokens(networkId))
+  }
+
   async function updateTokenDetails(networkId: number, numTokens: number): Promise<void> {
+    // Fetch addresses from TCR
+    const tcrAddressesSet = await retry(() => fetchTcrAddresses(networkId))
+    logDebug(`[tokenListFactory][${networkId}] TCR contains ${tcrAddressesSet.size} addresses`)
+
     // Fetch addresses from contract given numTokens count
     const addressesAndIds = await retry(() => fetchAddressesAndIds(networkId, numTokens))
```

Comment:
This line and 141 can be done in parallel with a Promise.all. 
They are slow IO operations, so it make sense to do it.